src/workerd/jsg/commonjs.c++

```diff
@@ -0,0 +1,163 @@
+    // If the specifier begins with one of our known prefixes, let's not resolve
+    // it against the referrer.
+    if (specifier.startsWith("node:") || specifier.startsWith("cloudflare:") ||
+        specifier.startsWith("workerd:")) {
+      return kj::Path::parse(specifier);
+    }
+    return path.parent().eval(specifier);
+  })();
+
+  // require() is only exposed to worker bundle modules so the resolve here is only
+  // permitted to require worker bundle or built-in modules. Internal modules are
+  // excluded.
+  auto& info = JSG_REQUIRE_NONNULL(modulesForResolveCallback->resolve(js, targetPath, path,
+                                       ModuleRegistry::ResolveOption::DEFAULT,
+                                       ModuleRegistry::ResolveMethod::REQUIRE, specifier.asPtr()),
+      Error, "No such module \"", targetPath.toString(), "\".");
+  // Adding imported from suffix here not necessary like it is for resolveCallback, since we have a
+  // js stack that will include the parent module's name and location of the failed require().
+
+  ModuleRegistry::RequireImplOptions options = ModuleRegistry::RequireImplOptions::DEFAULT;
+  if (getCommonJsExportDefault(js.v8Isolate)) {
+    options = ModuleRegistry::RequireImplOptions::EXPORT_DEFAULT;
+  }
+
+  return ModuleRegistry::requireImpl(js, info, options);
+}
+
+CommonJsModuleObject::CommonJsModuleObject(jsg::Lock& js)
+    : exports(js.v8Isolate, v8::Object::New(js.v8Isolate)) {}
+
+v8::Local<v8::Value> CommonJsModuleObject::getExports(jsg::Lock& js) {
+  return exports.getHandle(js);
+}
+void CommonJsModuleObject::setExports(jsg::Value value) {
+  exports = kj::mv(value);
+}
+
+void CommonJsModuleObject::visitForMemoryInfo(MemoryTracker& tracker) const {
+  tracker.trackField("exports", exports);
+}
+
+// ======================================================================================
+
+NodeJsModuleContext::NodeJsModuleContext(jsg::Lock& js, kj::Path path)
+    : module(jsg::alloc<NodeJsModuleObject>(js, path.toString(true))),
+      path(kj::mv(path)),
+      exports(js.v8Ref(module->getExports(js))) {}
+
+v8::Local<v8::Value> NodeJsModuleContext::require(jsg::Lock& js, kj::String specifier) {
+  // If it is a bare specifier known to be a Node.js built-in, then prefix the
+  // specifier with node:
+  bool isNodeBuiltin = false;
+  auto resolveOption = jsg::ModuleRegistry::ResolveOption::DEFAULT;
+  KJ_IF_SOME(spec, checkNodeSpecifier(specifier)) {
+    specifier = kj::mv(spec);
+    isNodeBuiltin = true;
+    resolveOption = jsg::ModuleRegistry::ResolveOption::BUILTIN_ONLY;
+  }
+
+  // TODO(cleanup): This implementation from here on is identical to the
+  // CommonJsModuleContext::require. We should consolidate these as the
+  // next step.
+
+  auto modulesForResolveCallback = jsg::getModulesForResolveCallback(js.v8Isolate);
+  KJ_REQUIRE(modulesForResolveCallback != nullptr, "didn't expect resolveCallback() now");
+
+  kj::Path targetPath = ([&] {
```

Comment:
This function is duplicate. We can reuse it if we move it outside.

Reply:
For this PR I'm intentionally just moving things around, not changing the current impl. I'm planning some other changes here that will come soon in a different PR